spirv_reflect.c

```diff
@@ -3332,6 +3494,282 @@ static SpvReflectResult ParseExecutionModes(
+    goto cleanup;                                                       \
+  }                                                                     \
+}
+
+#define SCALAR_TYPE_FLAGS (SPV_REFLECT_TYPE_FLAG_BOOL | SPV_REFLECT_TYPE_FLAG_INT | SPV_REFLECT_TYPE_FLAG_FLOAT)
+#define SCALAR_DISALLOWED_FLAGS (~0 ^ SCALAR_TYPE_FLAGS)
+#define VECTOR_TYPE_FLAGS (SCALAR_TYPE_FLAGS | SPV_REFLECT_TYPE_FLAG_VECTOR)
+#define VECTOR_DISALLOWED_FLAGS (~0 ^ VECTOR_TYPE_FLAGS)
+
+#define COMPOSITE_TYPE_FLAGS (VECTOR_TYPE_FLAGS|SPV_REFLECT_TYPE_FLAG_MATRIX|SPV_REFLECT_TYPE_FLAG_STRUCT|SPV_REFLECT_TYPE_FLAG_ARRAY)
+#define COMPOSITE_DISALLOWED_FLAGS (~0 ^ COMPOSITE_TYPE_FLAGS)
+
+static SpvReflectResult ParserGetScalarConstant(const SpvReflectShaderModule* p_module, SpvReflectPrvParser* p_parser, SpvReflectPrvNode* p_node,
+    SpvReflectScalarValueData* result, SpvReflectTypeDescription** type)
+{
+  SpvReflectTypeDescription* d_type = FindType(p_module, p_node->result_type_id);
+  if (!d_type) return SPV_REFLECT_RESULT_ERROR_SPIRV_INVALID_ID_REFERENCE;
+
+  if (d_type->type_flags & SCALAR_DISALLOWED_FLAGS) return SPV_REFLECT_RESULT_ERROR_SPIRV_INVALID_TYPE;
+  uint32_t low_word;
+  CHECKED_READU32(p_parser, p_node->word_offset + 3, low_word);
+  // There is no alignment requirements in c/cpp for unions
+  if (d_type->traits.numeric.scalar.width == 32) {
+    memcpy(&result->value.uint32_bool_value, &low_word, 4);
+  }
+  else if (d_type->traits.numeric.scalar.width == 64) {
+    uint32_t high_word;
+    CHECKED_READU32(p_parser, p_node->word_offset + 4, high_word);
+    uint64_t combined = low_word | (((uint64_t)high_word) << 32);
+    memcpy(&result->value.uint64_value, &combined, 8);
+  }
+  else {
+    return SPV_REFLECT_RESULT_ERROR_SPIRV_INVALID_TYPE;
+  }
+  *type = d_type;
+  return SPV_REFLECT_RESULT_SUCCESS;
+}
+
+#define IS_SPEC_LITERAL_OP(op_code)\
+    (((op_code) == SpvOpSpecConstantTrue) || ((op_code) == SpvOpSpecConstantFalse) || ((op_code) == SpvOpSpecConstant))
+#define IS_LITERAL_OP(op_code)\
+    (((op_code) == SpvOpConstantTrue) || ((op_code) == SpvOpConstantFalse) || ((op_code) == SpvOpConstant))
+#define IS_CONSTANT_LITERAL_OP(op_code) (IS_SPEC_LITERAL_OP(op_code)||IS_LITERAL_OP(op_code))
+// constant null and constant sampler etc are not handled in literals or anywhere for now...
+#define IS_CONSTANT_OP(op_code) \
+    (IS_CONSTANT_LITERAL_OP(op_code) || ((op_code) == SpvOpSpecConstantOp) || ((op_code) == SpvOpSpecConstantComposite) \
+    || ((op_code) == SpvOpConstantComposite) || ((op_code) == SpvOpConstantNull) || ((op_code) == SpvOpConstantSampler) \
+    || ((op_code) == SpvOpConstantPipeStorage) || ((op_code) == SpvOpSpecConstantCompositeContinuedINTEL)\
+    || ((op_code) == SpvOpConstantFunctionPointerINTEL) || ((op_code) == SpvOpConstantCompositeContinuedINTEL))
+
+#if SPIRV_REFLECT_ENABLE_CONSTANT_EVALUATION
+#ifdef _MSC_VER
+#define SPIRV_REFLECT_FORCEINLINE __forceinline
```

Comment:
FORCEINLINE can make the evaluation kernel function compile into multiple function bodies that have no overhead of inner switches. Advantage over writing multiple functions is that state tracking as here in two preprocessing passes and in evaluation is done in the same place. This can save a lot of going up and down through code. Easier to maintain.